### ISSUE TITLE
[rllib] Provide atari results across all algorithms (as applicable)

### ISSUE THREAD

**Issue by ericl** [ray-project/ray#2663]
### Describe the problem
We should publish results for at least a few of the standard Atari games on all applicable algorithms, and fix any discrepancies, e.g. https://github.com/ray-project/ray/issues/2654

Results uploaded to this repo: https://github.com/ray-project/rl-experiments

- [x] IMPALA
- [ ] IMPALA-LSTM
- [ ] A3C
- [x] A2C
- [ ] DQN
- [ ] APEX
- [x] PPO

Envs to run: PongNoFrameskip-v4, BreakoutNoFrameskip-v4, BeamRiderNoFrameskip-v4, QbertNoFrameskip-v4, SpaceInvadersNoFrameskip-v4

(Chosen such that all but pong can run concurrently on a g3.16xl machine).

Some references:
https://github.com/btaba/yarlp
https://github.com/openai/baselines/issues/176

**Comment by ericl** [comment]
Just ran a "30% full speed" IMPALA across a couple environments. The results are pretty reasonable at 40M frames, with Qbert / Space invaders about inline with results from the A3C paper, and Breakout / Beamrider a bit below. Note that the episode max reward for Breakout and Beamrider are pretty good, but the mean is not quite up there.

I'm guessing we can improve on this with some tuning.

```
# Runs on a single g3.16xl node
atari-impala:
    env:
        grid_search:
            - BreakoutNoFrameskip-v4
            - BeamRiderNoFrameskip-v4
            - QbertNoFrameskip-v4
            - SpaceInvadersNoFrameskip-v4 
    run: IMPALA
    config:
        sample_batch_size: 250  # 50 * num_envs_per_worker
        train_batch_size: 500
        num_workers: 12
        num_envs_per_worker: 5
```

![atari-impala](https://user-images.githubusercontent.com/14922/44184639-d4dbaa00-a0c4-11e8-8a31-4ebf2df3e829.png)

**Comment by robertnishihara** [comment]
In what format does it make sense to publish the results? E.g., a collection of full learning curves (e.g., as CSV)? Or actual visualizations like you have above? Or something else?

**Comment by ericl** [comment]
If we have a public ray perf dashboard, that would be a good place to put these.

Otherwise, I think posting some summary visualizations on github or the docs would do. The full learning curve data probably isn't that interesting, but we could also upload that to S3 pretty easily.

**Comment by luochao1024** [comment]
Do you have any result about A3C or A3C-LSTM?

**Comment by ericl** [comment]
I did an initial run with A3C, however the results were much worse as the
Impala ones. I didn't try tuning the learning rate though as mentioned in
the A3C paper.

**Comment by ericl** [comment]
That PR also adds A2C. Since A2C is deterministic, it should be easy to copy hyperparameters from another A2C implementation to compare results.

**Comment by luochao1024** [comment]
you are using 11 workers for experiment. I would recommend 16 workers.

**Comment by ericl** [comment]
@luochao1024 this PR reproduces standard Atari results for IMPALA and A2C: https://github.com/ray-project/ray/pull/2700

I'm still having trouble finding the right hyperparams for A3C, but since it works in A2C it's probably just a matter of tweaking the lr / batch size.

**Comment by luochao1024** [comment]
Do you have some right hyperparams that work for a3c now?

**Comment by luochao1024** [comment]
@ericl Can you give it a try for BreakoutNoFrameskip-v4? I try a grid search for the lr, but I still get some really bad results. Here is the configs I use:

```
atari-a3c:
    env: BreakoutNoFrameskip-v4
    run: A3C
    config:
        num_workers: 8
        sample_batch_size: 20
        use_pytorch: false
        vf_loss_coeff: 0.5
        entropy_coeff: -0.01
        gamma: 0.99
        grad_clip: 40.0
        lambda: 1.0
        lr:
            grid_search:
                - 0.000005
                - 0.00001
                - 0.00005
                - 0.0001
        observation_filter: NoFilter
        preprocessor_pref: rllib
        num_envs_per_workers: 5
        optimizer:
            grads_per_step: 1000
```

**Comment by ericl** [comment]
You'll definitely need to use the deepmind preprocessors, since the rllib
knees don't have the right episodic life wrappers. Perhaps we should remove
those. Also, maybe don't use LSTM and start from the A2C config.

**Comment by ericl** [comment]
There's this one weird thing where num_envs_per_worker will reduce your
effective unroll length per env (so 20 / 5 = unroll length of 4). So just
watch out for that and you might consider trying 1 env per worker instead,
or setting sample_batch_size=50 for a longer unroll.

Beyond that the config looks fine. Note that I found a lr schedule is
important for some envs (but it's probably too much to try right now).

**Comment by luochao1024** [comment]
The result seems normal now with num_workers=5

BreakoutNoFrameskip-v4:
![image](https://user-images.githubusercontent.com/38442610/44811191-9997ab80-ab99-11e8-8419-2066dc68a9ab.png)



QbertNoFrameskip-v4:
![image](https://user-images.githubusercontent.com/38442610/44811437-4e31cd00-ab9a-11e8-898f-a4876e4cfe3a.png)

I will set the num_envs_per_worker=1 later

**Comment by ericl** [comment]
Closing this in favor of individual tickets. Main TODOs are the DQN family.
